src/StringOrArrayDynamicFunctionReturnTypeExtension.php

```diff
@@ -29,21 +35,63 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
     // phpcs:ignore SlevomatCodingStandard.Functions.UnusedParameter
     public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
     {
-        $args = $functionCall->getArgs();
-        if (count($args) === 0) {
+        if (count($functionCall->getArgs()) === 0) {
             return null;
         }
 
-        $dataArgType = $scope->getType($args[0]->value);
-        if ($dataArgType->isArray()->yes()) {
-            $keyType = $dataArgType->getIterableKeyType();
-            if ($keyType instanceof StringType) {
-                return new ArrayType(new StringType(), new StringType());
+        return TypeTraverser::map(
+            $scope->getType($functionCall->getArgs()[0]->value),
+            function (Type $type, callable $traverse): Type {
+                if ($type instanceof UnionType || $type instanceof IntersectionType) {
+                    // phpcs:ignore NeutronStandard.Functions.VariableFunctions.VariableFunction
+                    return $traverse($type);
+                }
+
+                return $this->getType($type);
             }
+        );
+    }
+
+    private function getType(Type $type): Type
+    {
+        if ($type->isScalar()->yes()) {
+            return new StringType();
+        }
+
+        if (!$type->isArray()->yes()) {
+            return new ConstantStringType('');
+        }
+
+        if (count($type->getConstantArrays()) > 0) {
+            return TypeCombinator::union(
+                ...array_map(
+                    function (ConstantArrayType $constantArray): Type {
+                        $builder = ConstantArrayTypeBuilder::createEmpty();
+                        foreach ($constantArray->getKeyTypes() as $i => $keyType) {
+                            $builder->setOffsetValueType(
+                                $keyType,
+                                $this->getType($constantArray->getValueTypes()[$i]),
+                                $constantArray->isOptionalKey($i)
+                            );
+                        }
 
-            return new ArrayType(new IntegerType(), new StringType());
+                        return $builder->getArray();
+                    },
+                    $type->getConstantArrays()
+                )
+            );
         }
 
-        return new StringType();
+        return TypeCombinator::union(
+            ...array_map(
+                function (ArrayType $arrayType): Type {
```

Comment:
```suggestion
                static function (ArrayType $arrayType): Type {
```

Reply:
this is not possible because it uses recursion in the callable

Reply:
Oh  noooooooooooooooooooooooooooooooooooooooooooo!